### PR TITLE
Improving Oracle query metrics algorithm

### PR DESCRIPTION
**Pull request by nenadnoveljic**
### What does this PR do?

This PR:

1. improves the Oracle query metrics algorithm. We are generating query metrics for many more queries than before.
2. The execution time of fetching is reduced by the factor of 1000.

#### Query metrics
The old query metrics algorithm didn't collect any query metrics on idle databases. 

#### Execution plan performance
Before this change, it took ~4s to fetch a single execution plan on a test database. Now the same operation lasts ~4ms.

#### Query metrics
The old algorithm calculated query metrics only for the queries which were captured in two consecutive snapshots. The primary goal was not to hurt the performance. We switched to the scanning algorithm, where we take the snasphots of the recently executed query in SQL cache (a.k.a shared pool) and calculation the difference to get the execution statistics in the time interval.

A sample output on the idle database - agent queries & white noise by some background processes:
image

#### Execution plan performance
`SQL_ID` was added to the filter to enable the indexed path.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: releasenotes/notes/oracle-query-metrics-improvement-9e23624f76923a74.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# CHANGELOG.rst is rendered. So the text needs to be worded so that
+# it does not depend on any information only available in another
+# section. This may mean repeating some details, but each section
+# must be readable independently of the other.
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    Improve Oracle query metrics algorithm and the execution time of execution plan fetching.
```
Comment (buraizu) [inline releasenotes/notes/oracle-query-metrics-improvement-9e23624f76923a74.yaml:11]:
```suggestion
    Improve Oracle query metrics algorithm and the fetching time for execution plans.
```

**Review by buraizu — APPROVED** [review]
Approving with an optional suggestion to reduce repetition of `execution`

File: pkg/collector/corechecks/oracle-dbm/oracle.go [hunk 2]
```diff
@@ -22,6 +22,7 @@ import (
 	"github.com/DataDog/datadog-agent/pkg/util/log"
 	"github.com/DataDog/datadog-agent/pkg/version"
 	_ "github.com/godror/godror"
+	ttlcache "github.com/jellydator/ttlcache/v2"
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/oracle.go:25]: Do you think you could use `github.com/patrickmn/go-cache` instead? Not that I think it's better than this one but it's already in the agent's dependencies and it probably makes sense to try that one before introducing something new.

File: pkg/collector/corechecks/oracle-dbm/oracle.go [hunk 3]
```diff
@@ -283,6 +295,16 @@ func (c *Check) Configure(integrationConfigDigest uint64, rawInstance integratio
...
 	c.tagsString = strings.Join(c.tags, ",")
+
+	c.fqtCache = ttlcache.NewCache()
+	c.fqtCache.SetTTL(time.Duration(60 * time.Minute))
+	c.planCache = ttlcache.NewCache()
+	var planCacheRetention = c.config.QueryMetrics.PlanCacheRetention
+	if planCacheRetention == 0 {
+		planCacheRetention = 1
+	}
+	c.planCache.SetTTL(time.Duration(time.Duration(planCacheRetention) * time.Minute))
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/oracle.go:306]: I think you can skip one type conversion here:
```suggestion
	c.planCache.SetTTL(time.Duration(planCacheRetention) * time.Minute)
```

File: pkg/collector/corechecks/oracle-dbm/statements.go [hunk 4]
```diff
@@ -388,33 +399,13 @@ type PlanRows struct {
...
-		err = db.Select(&statementMetrics, db.Rebind(query), args...)
-		if err != nil {
-			return nil, fmt.Errorf("error executing statement metrics query: %w %s", err, statements_metrics_query)
-		}
-		return statementMetrics, nil
+func GetStatementsMetricsForKeys(c *Check, key string, negator string) ([]StatementMetricsDB, error) {
+	var statementMetrics []StatementMetricsDB
+	err := selectWrapper(c, &statementMetrics, fmt.Sprintf(STATEMENT_METRICS_QUERY, key, negator, key), 2*c.config.QueryMetrics.CollectionInterval, c.config.QueryMetrics.DBRowsLimit)
+	if err != nil {
+		return nil, fmt.Errorf("error executing statement metrics query: %w %s", err)
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/statements.go:406]: You've got an extra `%s` without a matching parameter to `Errorf`.

File: pkg/collector/corechecks/oracle-dbm/statements.go [hunk 5]
```diff
@@ -957,10 +847,29 @@ func (c *Check) StatementMetrics() (int, error) {
 	c.statementsCache.SQLIDs = nil
 	c.statementsCache.forceMatchingSignatures = nil
 
-	c.statementsLastRun = start
-
-	if SQLTextErrors > 0 || planErrors > 0 {
-		return SQLCount, fmt.Errorf("SQL statements processed: %d, text errors: %d, plan erros: %d", SQLCount, SQLTextErrors, planErrors)
+	if planErrors > 0 {
+		return SQLCount, fmt.Errorf("SQL statements processed: %d, plan erros: %d", SQLCount, planErrors)
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/statements.go:851]:
```suggestion
		return SQLCount, fmt.Errorf("SQL statements processed: %d, plan errors: %d", SQLCount, planErrors)
```

File: pkg/collector/corechecks/oracle-dbm/oracle.go [hunk 6]
```diff
@@ -283,6 +295,16 @@ func (c *Check) Configure(integrationConfigDigest uint64, rawInstance integratio
 	c.tags = append(c.tags, fmt.Sprintf("dbms:%s", common.IntegrationName), fmt.Sprintf("ddagentversion:%s", c.agentVersion))
 
 	c.tagsString = strings.Join(c.tags, ",")
+
+	c.fqtCache = ttlcache.NewCache()
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/oracle.go:299]: I'm not sure what the proper name would be for those cache instances but reading `fqtCache`, my intuition would be that we're caching the fqt (or that we're caching plans for the `planCache`) and not the fact that it was already sent during the cache's ttl.

File: pkg/collector/corechecks/oracle-dbm/oracle.go (hunk 2 above)
Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/oracle.go]: The Agent uses `github.com/patrickmn/go-cache` now.

File: pkg/collector/corechecks/oracle-dbm/oracle.go (hunk 6 above)
Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/oracle.go]: Renamed to `fqtEmitted`.

File: go.mod [hunk 7]
```diff
@@ -571,6 +571,8 @@ require (
 	github.com/aquasecurity/trivy v0.0.0-00010101000000-000000000000
 	github.com/cloudfoundry-community/go-cfclient/v2 v2.0.1-0.20230503155151-3d15366c5820
 	github.com/godror/godror v0.37.0
+	github.com/jellydator/ttlcache/v2 v2.11.1
```
Comment (alexandre-normand) [inline go.mod:574]: I think you can now run `go mod tidy` to clean up the ttlcache dependency that's unused.

Comment (nenadnoveljic) [reply go.mod]: done, thanks!

File: pkg/collector/corechecks/oracle-dbm/statements.go [hunk 8]
```diff
@@ -597,99 +562,32 @@ func (c *Check) StatementMetrics() (int, error) {
...
-						queryRow.Commands = obfuscatedStatement.commands
-						queryRow.QuerySignature = obfuscatedStatement.querySignature
-						queryRow.Tables = obfuscatedStatement.tables
-						SQLStatement = obfuscatedStatement.statement
-
-						c.DDstatementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature] = c.DDPrevStatementsCache.forceMatchingSignatures[statementMetricRow.StatementMetricsKeyDB.ForceMatchingSignature]
-						found = true
-					}
+			if statementMetricRow.SQLTextLength == 1000 {
+				getFullSQLText(c, &SQLStatement, "sql_id", statementMetricRow.RandomSQLID)
```
Comment (alexandre-normand) [inline pkg/collector/corechecks/oracle-dbm/statements.go:569]: The linter won't let you skip checking the error returned here. But I think it's also a good idea to handle the error.

Comment (nenadnoveljic) [reply pkg/collector/corechecks/oracle-dbm/statements.go]: fixed now, thx/

**Review by alexandre-normand — APPROVED** [review]

**Pull request merged**